### PR TITLE
WIP: Feature/offline

### PR DESCRIPTION
**Pull request by imtoori**
Offline storage works using the library moor (https://moor.simonbinder.eu), a sqlite orm.

The class `OfflineStorage` manages the database synchronization using a background isolate in order to unblock the main thread.
The `Client` uses `OfflineStorage` to synchronize the database with the newest information every time it receives new data about channels/messages.

### Multi-user
The db file is named after the userId, so if you instantiate a client using a different userId you will use different databases.
Calling `client.disconnect` flushes all current database data.

## Db structure
There are 7 tables:
 
- ChannelQueries
- Channels
- Users
- Reads
- Reactions
- Messages
- Members

Tables are normalized.

## Query channel mapping
The table `ChannelQueries` links a channel to a query.
Every row stores a `channelCid` and a `queryHash` derived by the filter options passed in `Channel.queryChannel` calls.
Everytime `Channel.queryChannel` is called the client

- retrieves the list of channel from the offline storage
- performs the api call

if the call resolves with no errors the channel list is updated with the fresh data and it inserts a new row for each channel in `ChannelQueries`.

If the pagination offset is 0 it also deletes every row related to the query before inserting the new rows.

### Sort and pagination
Sort and pagination happen during database queries.

## Updating/deleting/sending a message while offline
The information about the action is saved in the offline storage.
When the client returns online everything is retried.

## Recover/go back online
We use `Client.queryChannel` to get the latest information about the channel in memory.

## Enable/disable offline storage
By default the offline storage is active.
There is a named parameter in the `Client` constructor called `persistenceEnabled`; set it to `false` to disable the storage.

## Encryption
At the moment there is no encryption.
There is the possibility to add it (https://moor.simonbinder.eu/docs/other-engines/encryption/), but it is a bit tricky. It will be reconsidered later.

### CODE REVIEW

**Comment by winwisely99** [comment]
I dont mean to be rude, but i really feel that using HIVE is a way better approach than use SQLLite.

Docs. https://docs.hivedb.dev
Code: https://github.com/hivedb/hive

1. Hive is way faster. Is about 50 times faster than SQLite.
  - https://github.com/hivedb/hive#benchmark
2. HIVE tolerates the Server and Client using different versions of the Data. Its basically designed to support Schema Evolution. SO if you app is deployed to IOS and is V1, but your Server is V2 and the types structure has change the IOS app will not crash. It will do exactly like what Protobufs do any elegantly ignore the fields it does not know about until it does get updates from our Overlords at Apple to V2.
3. Hive runs on web, desktop and mobiel with zero native dependencies.
4. Data Model mismatch: https://docs.hivedb.dev/#/best-practices/when_to_use_hive?id=when-not-to-use-hive
  - basically for chat a KV like hive is a create match. you dont need SQL for essentially buckets.
5. Hive has encryption at rest. you can use AES and there are packages for Flutter that extend it to use different crypto also.
  - For example: https://pub.dev/packages/steel_crypt
6. Hive has migrations also.
7. Hive can use GRPC and Prootbuffers. We do. Its fast and solves the client and server beng at different versions and was designed to support Schema Evolution from day one. With just JSON your going to find your in big trouble with upgrading your servers and supporting so many different clients ( swift, flutter, etc etc etc).

**Comment by tbarbugli** [comment]
Thank you for the input, that's never rude :)

Regarding the benchmark, I had a look at how the comparison is done and it looks like the batch operations for Sqlite are written in a very inefficient way (read one row inside a for-loop). In real-life you would optimize reads/writes into something like this:

```sql
SELECT * FROM messages WHERE id IN (...)
```

@imtoori my gut feeling is that we are better off using rock solid sqlite than something else. what do you think?

> I dont mean to be rude, but i really feel that using HIVE is a way better approach than use SQLLite.
> 
> Docs. https://docs.hivedb.dev
> Code: https://github.com/hivedb/hive
> 
> 1. Hive is way faster. Is about 50 times faster than SQLite.
> 
> * https://github.com/hivedb/hive#benchmark
> 
> 1. HIVE tolerates the Server and Client using different versions of the Data. Its basically designed to support Schema Evolution. SO if you app is deployed to IOS and is V1, but your Server is V2 and the types structure has change the IOS app will not crash. It will do exactly like what Protobufs do any elegantly ignore the fields it does not know about until it does get updates from our Overlords at Apple to V2.
> 2. Hive runs on web, desktop and mobiel with zero native dependencies.
> 3. Data Model mismatch: https://docs.hivedb.dev/#/best-practices/when_to_use_hive?id=when-not-to-use-hive
> 
> * basically for chat a KV like hive is a great match. You dont need SQL for essentially buckets.
> 
> 1. Hive has encryption at rest. You can use AES and there are packages for Flutter that extend it to use different crypto also.
> 
> * For example: https://pub.dev/packages/steel_crypt
> 
> 1. Hive has migrations also.
> 2. Hive can use GRPC and Protobuffers. We do. Its fast and solves the client and server beng at different versions and was designed to support Schema Evolution from day one. With just JSON your going to find your in big trouble with upgrading your servers and supporting so many different clients ( swift, flutter, etc etc etc).

**Comment by imtoori** [comment]
Thanks for the feedback, but in my personal opinion picking SQLite is the best choice for now.
I tried Hive before choosing Moore, but even model relations are simpler in Sqlite, at least for our use case.

**Comment by imtoori** [comment]
Closing for #17
